pep-9999.rst

```diff
@@ -0,0 +1,450 @@
+most importantly including run-time checks that guarantee that
+an immutable object does not change.
+
+Close prior work if `PEP 351: The freeze
+protocol <https://peps.python.org/pep-0351/>`_ which
+created immutable *copies* of mutable objects. This PEP
+is the spiritual ancestor to this PEP which tackles the
+probles of the ancestor PEP and more (e.g. meaning of
+immutability when types are mutable, immortality, etc).
+
+Immutability in action:
+
+.. code-block:: python
+
+	class Foo: pass
+
+	f = Foo()
+	g = Foo()
+	h = Foo()
+
+	f.f = g
+	g.f = h
+	h.f = g # cycle!
+
+	g.x = "African Swallow" # OK
+	freeze(f) # Makes, f, g and h immutable
+	g.x = "European Swallow" # Throws immutability exception
+
+
+Motivation
+==========
+
+
+Ensuring Data Integrity
+-----------------------
+
+Python programs frequently manipulate large, interconnected data
+structures such as dictionaries, lists, and user-defined objects.
+Unintentional mutations can introduce subtle and
+difficult-to-debug errors. By allowing developers to explicitly
+freeze objects and their transitive dependencies, Python can
+provide stronger correctness guarantees for data processing
+pipelines, functional programming paradigms, and API boundaries
+where immutability is beneficial.
+
+
+Eliminating Data Races in Concurrent Code
+-----------------------------------------
+
+Python’s Global Interpreter Lock (GIL) mitigates many data race
+issues, but as Python evolves towards improved multi-threading and
+parallel execution (e.g., subinterpreters and free-threaded Python
+efforts), data races on shared mutable objects become a more
+pressing concern. A deep immutability mechanism ensures that
+shared objects are not modified concurrently, enabling safer
+multi-threaded and parallel computation.
+
+See also the discussion about extensions further down in this
+document.
+
+
+Optimisations and Caching Benefits
+----------------------------------
+
+Immutable objects provide opportunities for optimisation, such as
+structural sharing, memoization, and just-in-time (JIT)
+compilation techniques (specialising for immutable data, e.g.
+fixed shape, fewer barriers, inlining, etc.). Freezing objects can
+allow Python to implement more efficient caching mechanisms and
+enable compiler optimisations that rely on immutability
+assumptions.
+
+
+Note that none of the above are *guaranteed* with immortal
+objects as these may still change if a program manipulates an
+object without knowing that it is immortal.
+
+
+Specification
+=============
+
+Changes to Python objects
+-------------------------
+
+Every Python object will have a flag that keeps track of its
+immutability status. Details about the default value of
+this flag is discussed further down in this document. 
+
+The single bit that is needed to store the flag can be stolen from
+**TODO** and hence object sizes are not changed by this PEP.
+
+**TODO** show where below the bit stealing happens.
```

Comment:
I would simplify this section in big part to prevent bike shedding. I already talked to Tobias offline about this. It would probably be sufficient to say that the object will not grow in size since we can steal a bit from XYZ. (With a bit more fluff) That should prevent bike shedding on this point.